### ISSUE TITLE
HTTPAuthSessionWrapper does not render guarded child resources

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4014]
|<img alt="unshift's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| unshift reported|
|-|-|
|Trac ID|trac#4014|
|Type|defect|
|Created|2009-09-11 04:25:22Z|
|Branch|https://github.com/twisted/twisted/tree/httpauth-render-4014|

When using twisted.guard.HTTPAuthSessionWrapper to guard child nodes of my Site object, rendering raises a NotImplementedException instead of calling the render method of the protected resource.

For more information, see: http://twistedmatrix.com/pipermail/twisted-web/2009-July/004215.html

My code in particular is something like:

```
    ...
    siteRoot = Resource()
    siteRoot.isLeaf = False

    
    jobWrapper = guard.HTTPAuthSessionWrapper(Portal(JobRealm(), [JobDBChecker(db)]), [guard.BasicCredentialFactory("foo realm")])
    siteRoot.putChild("job", jobWrapper)
```

However POSTs and GETs to /job result in NotImplementedExceptions, because the request is handled in HTTPAuthSessionWrapper and not passed through to JobRealm's avatar.

According to an IRC conversation this should work if the root node at /job has no render methods and I only call /job/_child_ entities, though I haven't confirmed this.

Attachments:

* [4014.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/144/144a4ff8db24bfa94498de6817d25db065020d0a/dc50b41f834507bd6c1f60a7c5735d964f5f85fb.py) (1790 bytes) - added by unshift on 2009-09-12 19:29:02Z - this should reproduce the bug

<details><summary>Searchable metadata</summary>

```
trac-id__4014 4014
type__defect defect
reporter__unshift unshift
priority__normal normal
milestone__ 
branch__branches_httpauth_render_4014 branches/httpauth-render-4014
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1252643122000000 1252643122000000
changetime__1253514945000000 1253514945000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [27322]) Branching to 'httpauth-render-4014'

**Comment by twisted-trac** [comment]
|<img alt="unshift's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|unshift commented|
|-|-|

Changing the title here to be a little more accurate

**Comment by twisted-trac** [comment]
|<img alt="unshift's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|unshift commented|
|-|-|

I think this depends on [#4018](https://github.com/twisted/twisted/issues/4018).  Even if HTTPAuthSessionWrapper's render() gets login code similar to that in getChildWithDefault and is able to render resources, traversing a sub-tree (as in my example code) is going to be problematic.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun removed owner|
|-|-|

Alright.  Rendering the avatar itself is now supported in the branch.  Please review.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve set owner to @exarkun|
|-|-|

Cool, +1!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set owner to @therve|
|-|-|

Thanks.  Can you do me a favor and merge it for me?

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve set status to `closed`|
|-|-|

(In [27353]) Merge httpauth-render-4014

Author: exarkun
Reviewer: therve
Fixes [#4014](https://github.com/twisted/twisted/issues/4014)

Add a render implementation to HTTPAuthSessionWrapper to allow Avatars to be
rendered.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
